axum/src/extract/path/mod.rs

```diff
@@ -143,6 +143,52 @@ use std::{
 ///
 /// If the URI cannot be deserialized into the target type the request will be rejected and an
 /// error response will be returned. See [`customize-path-rejection`] for an example of how to customize that error.
+///
+/// # Optional path parameters
+///
+/// You can have a single handler for routes with different path parameters, but `Path` extractor
+/// will fail if a route does not match, so you would need to define separate routes with the same
+/// handler.
+///
+/// **Please note**: doing so with an `Option` wrapper is discouraged since the extractor may fail
+/// in other ways than because of a missing parameter, that is why you should use `Result` wrap
+/// and parse the resulting error:
+///
+/// ```rust,no_run
+/// use axum::{
+///     extract::{Path, rejection::PathRejection},
+///     routing::get,
+///     Router,
+/// };
+///
+/// async fn get_user(id: Result<Path<u128>, PathRejection>) {
+///     match parameter {
+///         Ok(Path(_)) => {
+///             // Route with a valid path parameter was used
+///         }
+///         Err(PathRejection::MissingPathParams(_)) => {
+///             // Route without a path parameter was used
+///         }
+///         Err(PathRejection::FailedToDeserializePathParams(_)) => {
+///             // Route with an invalid path parameter was used
+///         }
+///         Err(_) => {
+///             // `PathRejection` is marked `#[non_exhaustive]` so match must
+///             // include a catch-all case.
+///         }
+///     }
+/// }
+///
+/// let app = Router::new()
+///     .route("/users/:id", get(get_user))
+///     .route("/users", get(get_user));
+///
+/// # async {
+/// # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+/// # };
+/// ```
```

Comment:
I'm not sure we should show examples of code we _don't_ recommend users write 🤔

I think this is a nicer way of doing it

```rust
async fn with_params(Path(id): Path<u32>) -> Response {
    do_thing(Some(id)).await
}

async fn without_params() -> Response {
    do_thing(None).await
}

async fn do_thing(param: Option<u32>) -> Response { ... }

Router::new()
    .route("/:id", get(with_params))
    .route("/", get(without_params));
```

I'd be okay with including something like that in the docs for `Path`.

Reply:
I thought the point was not to use the `Option<Path>` rather than not to use `Result<Path>`. Maybe it would be OK to add both this to the example and `Result<Path>` as a manual handling example?

Reply:
I don't think `Result<Path>` (or `Option<Path>`) should be used to implement optional path params. So I don't think we should have an example that shows that. `Result<Path>` can be used to customize deserialization errors but thats already covered by the existing extractor docs.

Reply:
I see, it makes sense to rewrite this example then, will do